Bugfix/button loading issue in modal

it-vegard opened the pull request:
### Description

Fix edge-case issue with setting `loading={true}` on a [Button in a Modal on page load](https://nav-it.slack.com/archives/C7NE7A8UF/p1729215916213429). (Why would anyone do this?)

Button would get `width: 0` because the width is calculated before the button is rendered.
This also caused the spinner not to show, because we checked `widthOverride` instead of the `loading` prop.

### Component Checklist 📝

- [ ] JSDoc
- [ ] Examples
- [ ] Documentation
- [ ] Storybook
- [ ] Style mappings (`@navikt/core/css/config/_mappings.js`)
- [ ] Component tokens (`@navikt/core/css/tokens.json`)
- [ ] CSS class deprecations (`@navikt/aksel-stylelint/src/deprecations.ts`)
- [ ] Exports (`@navikt/core/react/src/index.ts` and `@navikt/core/react/package.json`)
- [ ] New component? CSS import (`@navikt/core/css/index.css`)
- [ ] Breaking change? Update migration guide. Consider codemod.
- [x] Changeset (Format: `<Component>: <gitmoji?> <Text>.` E.g. "Button: :sparkles: Add feature xyz.")

it-vegard committed:
Check loading prop for showing spinner, not widthOverride.

Making this change because widthOverride sometimes gets set to 0 (should also be fixed), meaning this check fails as "falsy".

it-vegard committed:
Prevent setting width to 0 while showing spinner, to avoid issue in Modals

Checking that width is truthy (not zero) will avoid issue where widthOverride gets set to 0 because the button is not yet rendered.

it-vegard committed:
Add changeset

KenAJoh commented on @navikt/core/react/src/button/Button.tsx:
Looks like width is not calculated correctly when set to loading by default 
<img width="322" alt="Screenshot 2024-10-18 at 10 28 56" src="https://github.com/user-attachments/assets/445bffa2-bf8a-4211-9fe9-0cf6b34a7027">

https://main--66b4b3beb91603ed0ab5c45e.chromatic.com/?path=/story/ds-react-button--loading
<img width="482" alt="Screenshot 2024-10-18 at 10 29 00" src="https://github.com/user-attachments/assets/791968da-dd63-4ba3-8125-676372c78a6f">

KenAJoh replied:
Is there any reason not to just set the text to invisible and absolute position center the loading 🤔 We might have overcomplicated this when calculating width.

it-vegard replied:
I have had a look at this, and currently see two solutions:

1) We revert line 135 back to check `widthOverride`. The reason setting the width fails, is that loading is set on first render, so the button never gets the larger size because of the label, while widthOverride is set after first render. In this case I recommend an inline comment to explain this.

2) The suggested change to only hide the label. We need e.g. `color: transparent` on both the text and icon, as well as `aria-hidden` to hide the label from screen readers, and for some reason need to change back to use `widthOverride` like above to have the exact same width. If not for that last detail, we get a few pixels difference, though I can't see why.

Option 1 is a lot simpler change, and does not require any ARIA. 
Option 2 is probably less fragile, but required both CSS and ARIA to work.

In any case, setting loading on page load should presumably be an edge case in the real world.

Any thought, @KenAJoh?

HalvorHaugan replied:
The pixel difference is probably caused by this CSS:
```css
.navds-button__icon:first-child {
  margin-left: var(--ac-button-icon-margin);
}
```
It can be fixed fixed by moving Loader between icon and label.

KenAJoh replied:
For nr.2: Is `visibility: hidden;` available for SR? It should (🤞🤞🤞) create no px shifts if both icon and text inside button changes visibility when loading

JulianNymark replied:
I also agree with the visibility hidden as the "simplest" solution here 🤔

it-vegard replied:
Yes, it seems like the best solution. I am just about to push that change.

it-vegard committed:
Replace width calculation with just visually hiding label + icon and absolutely positioning the loading icon.

it-vegard commented on @navikt/core/react/src/button/Button.tsx:
```diff
@@ -82,29 +81,14 @@ export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
-
       const filterProps: React.ButtonHTMLAttributes<HTMLButtonElement> =
-        disabled ?? widthOverride ? omit(rest, ["href"]) : rest;
+        disabled ?? loading ? omit(rest, ["href"]) : rest;
```
This line seems like it never worked as intended. Shouldn't it be as follows?

```suggestion
        (disabled || loading) ? omit(rest, ["href"]) : rest;
```

KenAJoh replied:
Nice catch 🐛

JulianNymark replied:
Indeed scary to have lots of boolean logic without parentheses 😅 (I try to always add too many as I never trust myself there, and in some cases the linter will remove the redundant ones).

it-vegard replied:
I think the reason TypeScript doesn't catch this is because none of the props are typed after going through `forwardRef`. Tried adding the type `<HTMLButtonElement, ButtonProps>` to `forwardRef`, but that did not accept the `as`-prop. Any ideas?

HalvorHaugan replied:
`OverridableComponent` creates some challenges with types inside the component. I don't really think there's a way to get "perfect" types as long as we accept any tag/component in `as`. You _could_ add this on line 81: `}: ButtonProps & { as?: React.ElementType },` but then you get issues when referring to `href`. I guess you could override that instance with `as any` though and still end up with improved types overall.

it-vegard committed:
Merge branch 'main' into bugfix/button-loading-issue-in-modal

HalvorHaugan commented on @navikt/core/css/button.css:
The CSS on line 112 (`.navds-button__icon:only-child`) needs to be modified to work when the loader is present.

Remember to also update button.darkside.css.

it-vegard replied:
Fixed with .navds-button--icon-only

it-vegard committed:
Update @navikt/core/react/src/button/Button.tsx

JulianNymark committed:
add story for bug case

JulianNymark committed:
Merge branch 'main' into bugfix/button-loading-issue-in-modal

KenAJoh commented on @navikt/core/css/button.css:
```diff
@@ -506,6 +510,10 @@
   stroke: var(--ac-button-primary-loader-stroke-bg, rgb(255 255 255 / 0.3));
 }
 
+.navds-button:has(.navds-loader) :is(.navds-button__icon, .navds-label) {
```
Since we have the class already, i guess we can just use it here instead of `:has`
```suggestion
.navds-button--loading :is(.navds-button__icon, .navds-label) {
```

KenAJoh replied:
Could it also make sense to do this? 🤔 
```
.navds-button--loading :is(:not(.navds-loader))
```

JulianNymark replied:
~~loading is a state right, so maybe this should use the data attrib too? 🤔~~ not important for this PR

it-vegard replied:
As long as we add a direct child descendant, that second suggestion works as well. Can't add a second suggestion in this thread, so I'll add it locally.

KenAJoh commented on @navikt/core/react/src/button/Button.tsx:
```diff
@@ -82,29 +81,14 @@ export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
       ref,
     ) => {
       const buttonRef = useRef<HTMLButtonElement | null>(null);
```
Cant see buttonRef or mergedRef being used anywhere now, can we just set `ref` directly on button and remove these? 🤩

it-vegard replied:
Good catch! Removed both.

HalvorHaugan reviewed (changes requested):
I dispute the approval since [this](https://github.com/navikt/aksel/pull/3252#discussion_r1822889232) hasn't been fixed yet 😅

KenAJoh commented on @navikt/core/css/button.css:
Icon-only variant changes width when switching betweens states
https://github.com/user-attachments/assets/9a975492-1bc3-4fce-bd1a-dff7d15d10f9

This happens since icon is no longer `:only-child` and loses some CSS while loading
<img width="469" alt="Screenshot 2024-12-19 at 15 00 48" src="https://github.com/user-attachments/assets/244f73db-6b9c-4a3a-87f3-fac9c2f72030" />


<img width="430" alt="Screenshot 2024-12-19 at 15 00 43" src="https://github.com/user-attachments/assets/c3b8408b-328d-4b58-a72e-c573589fbf71" />

it-vegard replied:
This is the same as [this](https://github.com/navikt/aksel/pull/3252#discussion_r1822889232) comment. Working on a fix for it.

it-vegard committed:
We no longer use buttonRef/mergedRef internally, so removed them and set ref directly

it-vegard committed:
Update @navikt/core/css/button.css

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Hiding everything except loading spinner is more robust than hiding every specific item.

it-vegard committed:
In loading-state, the button icon will no longer be able to be the only child (hidden), because of the loader icon. Instead checks for no label.

Could consider :not(:has(:is(.navds-button__icon, .navds-loader))) as well, to be more robust, but it might be overkill.

HalvorHaugan commented on @navikt/core/react/src/button/Button.tsx:
```diff
@@ -113,41 +92,33 @@ export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
-            width: widthOverride,
-          }}
-          disabled={disabled ?? widthOverride ? true : undefined}
+          style={style}
```
I think we can remove `style={style}` here, and remove `style` from the destructuring, so that it ends up in `rest`.

it-vegard replied:
Good catch!

it-vegard committed:
Much simpler/easier to read

it-vegard committed:
Don't need to destructure style, only to add it along with rest.

it-vegard commented on @navikt/core/react/src/button/Button.tsx:
```diff
@@ -113,41 +91,32 @@ export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
           {...filterProps}
-          ref={mergedRef}
+          ref={ref}
           onKeyUp={composeEventHandlers(filterProps.onKeyUp, handleKeyUp)}
```
Should we get `onKeyUp` from destructuring of the `props`, so it is not added through `filteredProps` and then overridden here?

HalvorHaugan replied:
Makes sense

HalvorHaugan commented on @navikt/core/css/button.css:
Now the only thing left is to update button.darkside.css as well 😁

it-vegard replied:
I was about to comment that. 🙈

it-vegard replied:
Added a quick edit for darkside. I'll welcome suggested improvements, e.g. for nesting choices.

it-vegard committed:
Should destructure onKeyUp from props, instead of having it added through filteredProps and then overridden afterwards.

JulianNymark committed:
Merge branch 'main' into bugfix/button-loading-issue-in-modal

JulianNymark reviewed (changes requested):
or wait, unapprove since darkside.css isn't done yet? (as mentioned)

it-vegard committed:
Quick and easy conversion for darkside

it-vegard committed:
Merge branch 'main' into bugfix/button-loading-issue-in-modal

HalvorHaugan commented on @navikt/core/css/darkside/button.darkside.css:
```diff
@@ -273,6 +277,10 @@
   stroke: rgb(255 255 255 / 0.3);
 }
 
+.navds-button--loading > :is(:not(.navds-loader)) {
```
Why use `:is` here?

it-vegard replied:
Den hang igjen fra tidligere. Fjernet både her og i button.css

it-vegard committed:
Unnecessary :is after refactoring

HalvorHaugan reviewed (approved):
LGTM

it-vegard committed:
Merge branch 'main' into bugfix/button-loading-issue-in-modal
